storage/repair.go

```diff
@@ -0,0 +1,278 @@
+
+	// Record size differences
+	sizeDiffSeries := diffRes.SizeDifferences.Series()
+	numBlocks := 0
+	for _, series := range sizeDiffSeries {
+		numBlocks += len(series.Blocks())
+	}
+	sizeDiffScope.Counter("repair.series").Inc(int64(len(sizeDiffSeries)))
+	sizeDiffScope.Counter("repair.blocks").Inc(int64(numBlocks))
+	r.logger.WithFields(
+		xlog.NewLogField("namespace", namespace),
+		xlog.NewLogField("shard", shard.ID()),
+		xlog.NewLogField("resultType", "sizeDiff"),
+	).Infof("numSeries=%d, numBlocks=%d", len(sizeDiffSeries), numBlocks)
+
+	// Record checksum differences
+	checksumDiffSeries := diffRes.ChecksumDifferences.Series()
+	numBlocks = 0
+	for _, series := range checksumDiffSeries {
+		numBlocks += len(series.Blocks())
+	}
+	checksumDiffScope.Counter("repair.series").Inc(int64(len(checksumDiffSeries)))
+	checksumDiffScope.Counter("repair.blocks").Inc(int64(numBlocks))
+	r.logger.WithFields(
+		xlog.NewLogField("namespace", namespace),
+		xlog.NewLogField("shard", shard.ID()),
+		xlog.NewLogField("resultType", "checksumDiff"),
+	).Infof("numSeries=%d, numBlocks=%d", len(checksumDiffSeries), numBlocks)
+}
+
+type repairFn func() error
+
+type dbRepairer struct {
+	sync.RWMutex
+
+	database      database
+	opts          Options
+	ropts         repair.Options
+	shardRepairer databaseShardRepairer
+
+	repairFn            repairFn
+	nowFn               clock.NowFn
+	logger              xlog.Logger
+	repairInterval      time.Duration
+	repairCheckInterval time.Duration
+	closed              bool
+}
+
+func newDatabaseRepairer(database database) (databaseRepairer, error) {
+	opts := database.Options()
+	ropts := opts.RepairOptions()
+	if ropts == nil {
+		return nil, errNoRepairOptions
+	}
+	if err := ropts.Validate(); err != nil {
+		return nil, err
+	}
+
+	r := &dbRepairer{
+		database:            database,
+		opts:                opts,
+		ropts:               ropts,
+		nowFn:               opts.ClockOptions().NowFn(),
+		logger:              opts.InstrumentOptions().Logger(),
+		repairInterval:      ropts.RepairInterval(),
+		repairCheckInterval: ropts.RepairCheckInterval(),
+	}
+	r.repairFn = r.Repair
+
+	return r, nil
+}
+
+func (r *dbRepairer) Start() {
+	if r.repairInterval <= 0 {
+		return
+	}
+
+	start := r.nowFn()
+	for {
+		r.RLock()
+		closed := r.closed
+		r.RUnlock()
+
+		if closed {
+			break
+		}
+
+		end := r.nowFn()
+		if end.Sub(start) < r.repairInterval {
+			time.Sleep(r.repairCheckInterval)
+			continue
+		}
+
+		if err := r.repairFn(); err != nil {
+			r.logger.Errorf("error repairing database: %v", err)
+		}
+
+		start = r.nowFn()
+	}
+}
+
+func (r *dbRepairer) Stop() {
+	r.Lock()
+	r.closed = true
+	r.Unlock()
+}
+
+func (r *dbRepairer) Repair() error {
+	// Don't attempt a repair if the database is not bootstrapped yet
+	if !r.database.IsBootstrapped() {
+		return nil
+	}
+
+	// Lazily create the shard repairer
+	if r.shardRepairer == nil {
```

Comment:
The same reason I'm not creating session upfront, to avoid the spurious connection errors. But if we are fine with those errors, there's no need for lazy creation.